### ISSUE TITLE
Track run-session support across Crabbox providers

### ISSUE THREAD

**Issue by zozo123** [openclaw/crabbox#182]
## Summary
Track adoption of Crabbox's generic `RunSessionHandle` / `FeatureRunSession` contract across providers so external orchestrators such as Harbor can acquire reusable leases in a provider-neutral way.

## Context
- Islo implementation PR: https://github.com/openclaw/crabbox/pull/181
- Harbor generic adapter PR: https://github.com/harbor-framework/harbor/pull/1745
- Harbor run-session consumer PR: https://github.com/harbor-framework/harbor/pull/1746

## Provider model
Providers that can create a reusable workspace should eventually be able to return a session handle from:

```sh
crabbox run --provider <provider> --keep --lease-output session.json -- true
```

The JSON should include at least provider, lease id, slug when available, reused/kept state, and cleanup command.

## Candidate follow-ups
- Audit current delegated providers for reusable session semantics.
- Add `FeatureRunSession` only where lifecycle and cleanup behavior are clear.
- Consider a provider-neutral session helper for exec/upload/download so callers do not need provider-specific SSH reconstruction.

**Comment by zozo123** [comment]
Status refresh: the linked Islo implementation has now merged: https://github.com/openclaw/crabbox/pull/181

This tracker should remain open for provider-by-provider adoption beyond Islo/Blacksmith. The remaining question is which other providers have clear reusable-session lifecycle and cleanup semantics before advertising `FeatureRunSession`.

@clawsweeper re-review

**Comment by steipete** [comment]
Maintainer status refresh on current `main` (`29b7e57`): the linked Islo work has landed, so the old review statement that only Blacksmith supports run-session is stale.

Current confirmed coverage:
- `blacksmith-testbox` / `blacksmith`: advertises `FeatureRunSession` and returns `RunSessionHandle` with provider, lease id, slug, reused/kept state, cleanup command, and Actions URL.
- `islo`: advertises `FeatureRunSession` and returns `RunSessionHandle` with provider, lease id, slug, reused/kept state, and cleanup command.

I would keep this tracker open for provider-by-provider adoption, but narrow the remaining decision to providers with clear reusable lifecycle semantics. The current delegated providers without `FeatureRunSession` are `railway`, `upstash-box`, `azure-dynamic-sessions`, `modal`, `wandb`, `cloudflare`, `e2b`, and `tensorlake`; these should not advertise the feature until acquire/reuse/keep/cleanup behavior is proven for each provider.

Suggested next step: audit those remaining delegated providers in small PRs, adding `FeatureRunSession` only where `crabbox run --provider <provider> --keep --lease-output session.json -- true` produces a usable cleanup handle and has real behavior proof.

**Comment by zozo123** [comment]
Audit complete. Posting back to steipete on the next step he requested.

---

@steipete Audit done. I went through every remaining delegated provider and checked each against the four run-session verbs (acquire / reuse / keep / cleanup) using the islo + recipe as the bar. I only mark a provider READY where all four are *proven in code* (concrete `Stop`/teardown, `shouldStop := acquired && !req.Keep`, `handleDelegatedRunFailure` already wired, stable lease id). Summary first, evidence below.

## Summary

| Provider | Verdict | One-line gap |
|---|---|---|
| `cloudflare` | READY | Lifecycle fully wired + tested; only missing the feature flag + `Session` construction + 2 tests. |
| `e2b` | READY | Full acquire/reuse/keep/cleanup exists; wire flag + `Session` on all paths + cleanup helper + 2 tests. |
| `modal` | READY | Persistent detached sandbox + Stop exist; wire flag + `Session` + `modalCleanupCommand` + 2 tests. |
| `wandb` | READY | Lifecycle matches islo line-for-line; `Features` is `nil` today, add flag + `Session` (cleanup uses `--id`). |
| `upstash-box` | READY | Box create/reattach/Stop all present; wire flag + `Session` + cleanup helper + 2 tests. |
| `azure-dynamic-sessions` | NEEDS-WORK | Lifecycle test-proven, but **no** `Session` wiring at all yet (flag + handle + helper + tests). |
| `tensorlake` | NEEDS-WORK | Lifecycle is an islo near-twin, but `Features` is empty and `Session` never set; also a stale `bridge.go` comment to fix. |
| `railway` | NOT-FEASIBLE | Service-control, not sandbox-provisioning: `Run` can't execute the command, `--keep`/`--reclaim` are rejected. |

Note on READY vs NEEDS-WORK: the split is deliberately about how much is left, not about lifecycle quality. The five READY providers already have at least one lifecycle-bearing return path that *almost* carries a handle, or a near-mechanical wiring delta. The two NEEDS-WORK providers have an equally real lifecycle but **zero** `Session` plumbing today, so they need the full flag + handle + helper + both tests — slightly more surface, same shape. None of READY/NEEDS-WORK requires new lifecycle behavior; all of it is wiring + tests.

## READY — small per-provider PR now

All five share the same shape (mirror islo `backend.go:231-241` + `isloCleanupCommand` `backend.go:737-739`): add `core.FeatureRunSession` to `Spec().Features`, build `result.Session` right after `shouldStop` is computed, carry it on **every** return path (including the bare early-error `RunResult{}` returns), re-stamp `result.Session.Kept = !shouldStop` after each `handleDelegatedRunFailure`, add a cleanup helper using `shellQuote`, and add the two tests (spec + behavioral handle incl. byte-exact `CleanupCommand`, on success and kept-after-failure).

- **`cloudflare`** — cleanest candidate. Acquire `backend.go:101-109`, reuse `resolveSandboxID` `backend.go:425-443`, `shouldStop := acquired && !req.Keep` `backend.go:118`, `handleDelegatedRunFailure` already called `backend.go:213,217`, teardown `Stop()` `backend.go:304-325` → `destroySandbox` (`DELETE /v1/sandboxes/{id}`, `client.go:161-163`). Create+exec+destroy already exercised end-to-end by `TestCloudflareRunCleanupDestroyUsesBoundedContext` (`backend_test.go:689`). Feature is unadvertised (`provider.go:28` = `{FeatureArchiveSync, FeatureCleanup}`) and `Session` is set nowhere. Recommend matching cloudflaredynamicworkers' `RunID == leaseID` convention — confirm.
- **`e2b`** — acquire `createSandbox` `backend.go:320-364`, reuse `resolveSandboxID` `backend.go:372-418`, `shouldStop` `backend.go:140`, failure hook `backend.go:235,239`, teardown `Stop()` `backend.go:303-318`. Exact cleanup string `crabbox stop --provider e2b --id <id>` already in code/tests (`backend.go:357`, `backend_test.go:461`). Subtlety: a reused raw-sandbox id yields a synthetic `e2b_<id>` lease, so the cleanup command must `shellQuote` whatever lease id is chosen so it round-trips through `resolveSandboxID`. Feature unadvertised (`provider.go:24` = `{FeatureURLBridge}`).
- **`modal`** — persistent sandbox via `createSandbox` → `client.CreateSandbox` with `sb.detach()` so it survives the process (`client.go:325-334`); reuse `resolveSandboxID` (claim / `cbx_` lease / raw id) `backend.go:330-377`; `shouldStop` `backend.go:85`; failure hook `backend.go:184,188`; teardown `Stop()` `backend.go:252-267` → `Terminate` `client.go:410-422`. `crabbox stop --provider modal --id <id>` already emitted at `backend.go:302`. Feature unadvertised (`provider.go:24`). Watch the pre-identity early errors `backend.go:58-83` — return before a sandbox exists, so guard or nil-Session those (verify against `run.go:1555-1562`).
- **`wandb`** — `Features` is currently `nil` (`provider.go:24`), so this is the only thing blocking it. Acquire `backend.go:55-69` (persistent CoreWeave sandbox kept alive by `sleep infinity`, `client.go:36-38`), reuse via `--id` skipping acquire+Stop (proven `TestWandbRunWithExistingIDSkipsAcquireAndStop` `backend_test.go:324-345`), `shouldStop` `backend.go:87`, failure hook `backend.go:137,141`, teardown `Stop()` `backend.go:203-213` → `client.go:426-451`. **Design call to confirm:** wandb's `Stop` only takes `--id <sandbox-id>` (`backend.go:204`), so its `CleanupCommand` must be `crabbox stop --provider wandb --id <id>`, unlike islo/docker which stop by bare lease id. Reuse caveat: env-var forwarding to an existing sandbox is rejected (`backend.go:74-81`), but that doesn't affect the session lifecycle.
- **`upstash-box`** — acquire `createBox` → `client.CreateBox` `client.go:89-132`, reuse `resolveBoxID` (claim / `cbx_` / box id / slug) `backend.go:356-394`, `shouldStop` `backend.go:90`, failure hook `backend.go:199,206`, teardown `Stop()` `backend.go:310-325` → `DeleteBoxes` + `removeLeaseClaim`. Stable lease from `boxNamePattern` `backend.go:455-466`. `RunResult` already carries `Provider/LeaseID/Slug` (`backend.go:171-173`) but never `Session`. Feature `{FeatureArchiveSync}` only (`provider.go:27`).

## NEEDS-WORK — lifecycle is real, run-session plumbing is entirely absent

These are not blocked on lifecycle; they're blocked on having *zero* `Session` wiring, so they're a touch more work than the READY set (full flag + handle + helper + both tests) but the same pattern.

- **`azure-dynamic-sessions`** — full lifecycle is test-proven: reuse without teardown `TestRunReusesClaimWithoutStoppingSession` `backend_test.go:113-136`, keep-on-failure `TestRunKeepOnFailureRetainsNewSession` `backend_test.go:85-111`; `shouldStop` byte-identical to islo `backend.go:89`; `Stop()` `backend.go:292-312` → `DeleteSession` (`POST /.management/stopSession`). But `provider.go:24` advertises only `{FeatureArchiveSync}` and `Run` sets `RunResult.Session` on no path (`backend.go:112,119,159`). Structurally identical to islo — sound candidate, just needs the full wiring landed in one commit.
- **`tensorlake`** — persistent Firecracker sandbox via `tensorlake sbx create` (`cli.go:105-140`), `tlsbx_<id>` lease (`backend.go:305`), reuse `resolveLeaseID` `backend.go:322-352`, `shouldStop` `backend.go:85`, failure hook `backend.go:162,166`, teardown `Stop()` `backend.go:277-292` → `tensorlake sbx terminate`. `Features` is empty `{}` (`provider.go:24`) and `Session` is never set. Extra cleanup item: the `bridge.go:10-14` comment claims Tensorlake is "a serverless function platform … not long-lived sandboxes," which contradicts the implemented create→exec→terminate model — stale, should be fixed/removed alongside the wiring.

## NOT-FEASIBLE — ephemeral / wrong model

- **`railway`** — do **not** add `FeatureRunSession`. It's `ProviderKindServiceControl` (`provider.go:22`) over a user-owned, out-of-band Railway service, not a crabbox-provisioned sandbox. `Run` unconditionally returns `exit(2)` for arbitrary commands (`backend.go:94`) — the `-- true` workload can never execute — and `rejectRailwayRunOptions` (`backend.go:85`, `427-460`) rejects the exact knobs run-session needs: `--keep` (`428-429`) and `--reclaim` (`431-433`). Warmup is deliberately hard-rejected to avoid leaking billable resources (`backend.go:73-81`). There is no acquired/shouldStop/lease identity to populate. `Stop()` exists (`backend.go:383-403`) but stops a deployment of the user's long-lived service, so embedding it as a `CleanupCommand` would hand orchestrators a misleading handle pointed at a resource crabbox doesn't own — precisely the dead-handle risk flagged in this issue. Correct as-is: `provider.go:24` = `{FeatureURLBridge}`. A run-session story here would require reversing the intentional Warmup rejection and inventing crabbox-owned provisioning — a separate design effort, not a wiring change.

## First PR is up: #352 (`cloudflare`)

Rather than just recommend, I went ahead and opened the first per-provider PR: **#352**. `cloudflare` was the lowest-risk READY candidate — the lease lifecycle is already exercised end-to-end by the existing httptest harness, so no live creds were needed for real behavior proof.

It advertises `FeatureRunSession`, builds a `RunSessionHandle` on every post-identity return path (re-stamping `Kept` after each `handleDelegatedRunFailure`), and uses the sibling `cloudflaredynamicworkers` cleanup convention `crabbox stop --provider cloudflare --id <leaseId>` with `RunID == leaseID`. Three tests cover spec advertisement, `--keep` (byte-exact cleanup command, `reused=false`/`kept=true`), and `--keep-on-failure` lease retention. `go build ./...`, `go vet`, and the cloudflare/all/cli test packages all pass.

One open question on #352: should `RunID == leaseID` for cloudflare, matching the sibling, or stay empty?

If that shape looks right, I'll send the other four READY providers (`e2b`, `modal`, `wandb`, `upstash-box`) as the same small per-provider PRs.
